### DISCUSSION TITLE
My migrations are not picking up the constraint names?

### DISCUSSION THREAD

**Discussion by ghost** [sqlalchemy/alembic#1194] [Usage Questions] [answered]
SQLAlchemy 2.0, Python 3.11, Alembic 1.10.

In my `env.py` I have:

```python
target_metadata = SQLModel.metadata

target_metadata.naming_convention = {
    "ix": "ix_%(column_0_label)s",
    "uq": "uq_%(table_name)s_%(column_0_N_name)s",
    "ck": "ck_%(table_name)s_%(constraint_name)s",
    "fk": "fk_%(table_name)s_%(column_0_N_name)s_%(referred_table_name)s",
    "pk": "pk_%(table_name)s_%(column_0_N_name)s",
}
```
... but when I `autogenerate` a migration (from a fresh DB instance), I see the following:

```python
# stuff elided
    op.create_table('models',
    sa.Column('model_name', sqlmodel.sql.sqltypes.AutoString(), nullable=False),
    sa.Column('description', sqlmodel.sql.sqltypes.AutoString(), nullable=False),
    sa.Column('created_at', sa.DateTime(timezone=True), server_default=sa.text('now()'), nullable=False),
    sa.Column('updated_at', sa.DateTime(timezone=True), server_default=sa.text('now()'), nullable=False),
    sa.PrimaryKeyConstraint('model_name'),
    schema='model_cache',
    )
# more stuff elided
```

I was expecting that PrimaryKeyConstraint to get a name, as defined in the `naming_convention` dict I pass in. Am I missing something, here?

TIA!

**Comment by zzzeek** [comment, marked as answer]
i believe that .naming_convention has to be there before the Table objects associated with the MetaData get set up.   As you build Table objects etc., the naming convention applies symbols to the "name" fields of all the Constraint objects that are what cause the names to happen.

this is why SQLAlchemy docs illustrate .naming_convention only set up as a constructor keyword argument on `MetaData`.  It has never been tested by just adding the convention after the fact like that (though should work as long as the Table objects are created after it's present).

**Reply by ghost** [reply]
This makes sense, thanks. All done.
